### PR TITLE
NIFI-6243 Add Support for AtomicDistributedCache to the HBase 1.x and…

### PR DESCRIPTION
**Pull request by shawnweeks**
… 2.x Map Cache Services

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message? 

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by shawnweeks** [comment]
This is not ready to merge, just getting the code out there so some other folks can start looking at it.

File: nifi-nar-bundles/nifi-standard-services/nifi-hbase_2-client-service-bundle/nifi-hbase_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_2_ClientService.java [hunk 1]
```diff
@@ -485,6 +486,22 @@ public boolean checkAndPut(final String tableName, final byte[] rowId, final byt
...
+        try (final Table table = connection.getTable(TableName.valueOf(tableName))) {
+            Put put = new Put(rowId);
+            put.addColumn(
+                    column.getColumnFamily(),
+                    column.getColumnQualifier(),
+                    column.getBuffer());
+            if ( value == null || timestamp == 0L ) {
+                return table.checkAndMutate(rowId, family).qualifier(qualifier).ifNotExists().thenPut(put);
+            } else {
+                return table.checkAndMutate(rowId, family).qualifier(qualifier).ifEquals(value).timeRange(TimeRange.at(timestamp)).thenPut(put);
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-standard-services/nifi-hbase_2-client-service-bundle/nifi-hbase_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_2_ClientService.java]: How does this `timeRange(TimeRange.at(timestamp))` affect the operation? I talked with my colleague who has better knowledge on HBase than me said it is used to get the persisted value at the specified timestamp to compare the value specified with `ifEquals(value)`. I haven't tested myself though.

I feel we can remove the `timeRange(TimeRange.at(timestamp))` part. If we don't specify timestamp, HBase will compare with the latest value.

Also, if HBase requires the original value for atomic mutation (and specifying timestamp doesn't add any benefit for NiFi use-case), I don't think we need to add this `checkAndPut` method. The existing `checkAndPut` would be sufficient.

How do you think?

Comment (shawnweeks) [reply nifi-nar-bundles/nifi-standard-services/nifi-hbase_2-client-service-bundle/nifi-hbase_2-client-service/src/main/java/org/apache/nifi/hbase/HBase_2_ClientService.java]: Agreed, I misunderstood what the timestamp was used for there and thought it was going to include the timestamp as part of the check.

**Comment by shawnweeks** [comment]
Reverted the changes to the HBase Services and Interface. At some point we probably want to convert the HBase 2.x client to use the new checkAndMutate API but checkAndPut works for now. Currently waiting for a local build to finish and I'll be amending this commit. Thanks for looking at this.

**Comment by shawnweeks** [comment]
Still need to create test cases.

**Comment by ijokarumawak** [comment]
Hi @shawnweeks 
Today, I had a chance to try this PR with a live HBase 2.0. I confirmed that HBase_2_ClientMapCacheService can be used as expected from Wait/Notify processors.
So, all we have to do is adding some unit test cases.

Will you be able to update TestHBase_2_ClientMapCacheService and TestHBase_1_1_2_ClientMapCacheService?

**Comment by shawnweeks** [comment]
I've been swamped at work the last couple of weeks but I should be able to do the test cases this week. Sorry for the delay.

**Comment by shawnweeks** [comment]
@ijokarumawak I added test cases for the anything I could think of. I had to refactor the Mock HBase Service because it wasn't behaving like HBase. It was storing data internally as a List<> instead of Map<> and wasn't allowing you to insert Rows again. If anyone knows why this approach may be bad let me know.

**Comment by shawnweeks** [comment]
I probably should refactor the Mock HBase Service more to eliminate all the loops through the rows since you can now just do a map.get()

**Comment by shawnweeks** [comment]
> I probably should refactor the Mock HBase Service more to eliminate all the loops through the rows since you can now just do a map.get()

Yeah nevermind. So much mocking here that I just broke things when I tried.

**Comment by ijokarumawak** [comment]
Thanks @shawnweeks for adding the tests! LGTM +1. Merging to master!

**Pull request closed without merging**